### PR TITLE
SOC-6210: Implement search groups by keyword

### PR DESCRIPTION
**Pull request by wkhessairi**
This PR adds the possibility of searching groups by a keyword. It returns groups which name contains the passed parameter "keyword"

### CODE REVIEW

File: component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/PatchedHibernateIdentityStoreImpl.java [hunk 1]
```diff
@@ -746,10 +748,18 @@ public IdentityObject findIdentityObject(IdentityStoreInvocationContext ctx, Str
...
-         StringBuilder hqlBuilderConditions = new StringBuilder(" where io.realm=:realm and io.identityType=:identityType");
+         StringBuilder hqlBuilderConditions = new StringBuilder(" where io.realm=:realm");
          queryParams.put("realm", realm);
-         queryParams.put("identityType", hibernateType);
-
+         /* BEGIN SOC-6210: Search for all groups by keyword. If type name passed is equals to @@ALL_GROUPS@@, then
+         exclude USER type (search for all group types) */
+         String hibernateTypeName = hibernateType.getName();
+         if (ALL_GROUPS_TYPE.equals(hibernateTypeName)) {
+            hqlBuilderConditions.append(" and io.identityType.id <> 1");
```
Comment (thomasdelhomenie) [inline component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/PatchedHibernateIdentityStoreImpl.java]: Don't we have a constant or a way to retrieve this id dynamically ? I am afraid it could be different depending on the configuration.

File: component/portal/src/main/java/org/exoplatform/portal/rest/GroupRestResourcesV1.java [hunk 2]
```diff
@@ -62,11 +63,8 @@ public Response getGroups(@Context UriInfo uriInfo,
     offset = offset > 0 ? offset : DEFAULT_OFFSET;
     limit = limit > 0 ? limit : DEFAULT_LIMIT;
 
-    ListAccess<Group> groups = groupSearchService.searchGroups(q);
-    limit = limit < groups.getSize() ? limit : groups.getSize();
-    Group [] allGroups = groups.load(offset, limit);
-    LinkedList<String> listAllGroups = new LinkedList(Arrays.asList(allGroups));
-    
+    Collection<Group> groups = groupSearchService.searchGroups(q, offset, limit);
+    LinkedList<String> listAllGroups = new LinkedList(Arrays.asList(groups.toArray()));
```
Comment (thomasdelhomenie) [inline component/portal/src/main/java/org/exoplatform/portal/rest/GroupRestResourcesV1.java]:
```suggestion
    List<String> listAllGroups = new LinkedList(Arrays.asList(groups.toArray()));
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java [hunk 3]
```diff
@@ -38,9 +40,11 @@
 
     public static final String GROUP_DESCRIPTION = "description";
 
+    public static final String ALL_GROUPS_TYPE = "@@ALL_GROUPS@@";
```
[the remark this answers is not in the document]
Comment (wkhessairi) [reply component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java]: yes

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java [hunk 4]
```diff
@@ -581,6 +585,16 @@ public Group findGroupById(String groupId) throws Exception {
...
     }
 
+    public ListAccess<Group> findGroupsByKeyword(String keyword) throws Exception {
+        if (log.isTraceEnabled()) {
+            Tools.logMethodIn(log, LogLevel.TRACE, "findGroupsByKeyword", null);
+        }
+
+        IdentitySearchCriteria identitySearchCriteria = new IdentitySearchCriteriaImpl().nameFilter("*" + keyword + "*");
+
+        return new IDMGroupListAccess(ALL_GROUPS_TYPE, identitySearchCriteria);
```
[the remark this answers is not in the document]
Comment (wkhessairi) [reply component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java]: In this method yes, we search all group types. But it can be useful for further use cases. You think that we should remove this parameter for now ?

File: component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/PatchedHibernateIdentityStoreImpl.java (hunk 1 above)
Comment (wkhessairi) [reply component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/PatchedHibernateIdentityStoreImpl.java]: Right. I will check

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/IDMGroupListAccess.java [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+        OrganizationService.class);
+  }
+
+  private List<org.picketlink.idm.api.Group> listQuery(int index, int length) throws Exception {
+    identitySearchCriteria.page(index, length);
+    return (List<org.picketlink.idm.api.Group>) getIDMService().getIdentitySession().
+        getPersistenceManager().findGroup(groupType, identitySearchCriteria);
+  }
+
+  private Group convertGroup(org.picketlink.idm.api.Group jbidGroup) throws Exception {
```
[the remark this answers is not in the document]
Comment (wkhessairi) [reply component/identity/src/main/java/org/exoplatform/services/organization/idm/IDMGroupListAccess.java]: Yes indeed. I will look on how to move this on some utils class

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (thomasdelhomenie) [reply component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java]: I don't think so. The notion of group type is not visible in our API, it is a internal concept of picketlink, so I would remove it yes. At worst we can add it later if needed.

[the remark this answers is not in the document]
Comment (wkhessairi) [reply component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java]: ok

File: component/identity/src/main/java/org/exoplatform/services/organization/search/GroupSearchService.java [hunk 6]
```diff
@@ -1,12 +1,13 @@
...
 
+import java.util.Collection;
+
 /**
  * Service to search in groups
  */
 public interface GroupSearchService {
 
-  ListAccess<Group> searchGroups(String term) throws Exception;
+  Collection<Group> searchGroups(String term, int offset, int limit) throws Exception;
```
[the remark this answers is not in the document]
Comment (wkhessairi) [reply component/identity/src/main/java/org/exoplatform/services/organization/search/GroupSearchService.java]: yes ok

File: component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/PatchedHibernateIdentityStoreImpl.java [hunk 7]
```diff
@@ -79,6 +79,8 @@
 
 import org.exoplatform.services.organization.idm.EntityMapperUtils;
 
+import static org.exoplatform.services.organization.idm.GroupDAOImpl.ALL_GROUPS_TYPE;
```
[the remark this answers is not in the document]
Comment (wkhessairi) [reply component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/PatchedHibernateIdentityStoreImpl.java]: ok

File: component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/PatchedHibernateIdentityStoreImpl.java (hunk 1 above)
Comment (wkhessairi) [reply component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/PatchedHibernateIdentityStoreImpl.java]: The only way that I found, is to test on [user object type](https://github.com/exodev/gatein-portal/blob/develop/component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/PatchedHibernateIdentityStoreImpl.java#L2936)

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java [hunk 8]
```diff
@@ -949,4 +962,56 @@ public String getPLIDMGroupName(String gtnGroupName) {
     public String getGtnGroupName(String plidmGroupName) {
         return orgService.getConfiguration().getGtnGroupName(plidmGroupName);
     }
+
+    public class IDMGroupListAccess implements ListAccess<Group>, Serializable {
```
Comment (thomasdelhomenie) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java]: Shouldn't we move it to its own class, like IDMUserListAccess ?

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java [hunk 9]
```diff
@@ -40,7 +45,7 @@
 
     private List<GroupEventListener> listeners_;
 
-    private static final String CYCLIC_ID = "org.gatein.portal.identity.LOOPED_GROUP_ID";
+    static final String CYCLIC_ID = "org.gatein.portal.identity.LOOPED_GROUP_ID";
```
Comment (thomasdelhomenie) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java]: Why did you remove "private" ?

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java (hunk 8 above)
Comment (wkhessairi) [reply component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java]: I tried with no success, since it uses many methods in common with GroupDAOImpl

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java (hunk 9 above)
Comment (wkhessairi) [reply component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java]: It is a mistake :)

File: component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/PatchedHibernateIdentityStoreImpl.java [hunk 10]
```diff
@@ -746,10 +748,18 @@ public IdentityObject findIdentityObject(IdentityStoreInvocationContext ctx, Str
...
-         StringBuilder hqlBuilderConditions = new StringBuilder(" where io.realm=:realm and io.identityType=:identityType");
+         StringBuilder hqlBuilderConditions = new StringBuilder(" where io.realm=:realm");
          queryParams.put("realm", realm);
-         queryParams.put("identityType", hibernateType);
-
+         /* BEGIN SOC-6210: Search for all groups by keyword. If type name passed is equals to @@ALL_GROUPS@@, then
+         exclude USER type (search for all group types) */
+         String hibernateTypeName = hibernateType.getName();
+         if (ALL_GROUPS_TYPE.equals(hibernateTypeName)) {
+            hqlBuilderConditions.append(" and lower(io.identityType.name) <> 'user'");
```
[the remark this answers is not in the document]
Comment (wkhessairi) [reply component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/PatchedHibernateIdentityStoreImpl.java]: Yes. right

File: component/identity/src/main/java/org/exoplatform/services/organization/search/GroupSearchServiceImpl.java [hunk 11]
```diff
@@ -20,17 +18,11 @@ public GroupSearchServiceImpl(OrganizationService organizationService) {
 
   @Override
   public ListAccess<Group> searchGroups(String keyword) throws Exception {
-    List<Group> groups = new LinkedList<>();
     if (StringUtils.isBlank(keyword)) {
-      groups.addAll(organizationService.getGroupHandler().getAllGroups());
+      return new ListAccessImpl<>(Group.class, new ArrayList<>(organizationService.getGroupHandler().getAllGroups()));
```
[the remark this answers is not in the document]
Comment (wkhessairi) [reply component/identity/src/main/java/org/exoplatform/services/organization/search/GroupSearchServiceImpl.java]: yes, I will update it

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java (hunk 8 above)
Comment (thomasdelhomenie) [reply component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java]: ok

**Comment by wkhessairi** [comment]
> @wkhessairi you should add tests is possible also.

It is already tested [here](https://github.com/exodev/core/pull/35/files#diff-62f25f7c63886367a43f283639ca75daR182).

File: component/identity/src/main/java/org/exoplatform/services/organization/search/GroupSearchServiceImpl.java [hunk 12]
```diff
@@ -1,8 +1,6 @@
 package org.exoplatform.services.organization.search;
 
-import java.util.LinkedList;
-import java.util.List;
-
+import java.util.ArrayList;
```
Comment (thomasdelhomenie) [inline component/identity/src/main/java/org/exoplatform/services/organization/search/GroupSearchServiceImpl.java]: Useless import
